File: core/src/main/java/com/whylabs/logging/core/ColumnProfile.java
package com.whylabs.logging.core;

import static com.whylabs.logging.core.SummaryConverters.fromSchemaTracker;
import static java.util.stream.Collectors.toSet;

import com.whylabs.logging.core.data.ColumnSummary;
import com.whylabs.logging.core.data.InferredType.Type;
import com.whylabs.logging.core.format.ColumnMessage;
import com.whylabs.logging.core.statistics.CountersTracker;
import com.whylabs.logging.core.statistics.NumberTracker;
import com.whylabs.logging.core.statistics.SchemaTracker;
import com.whylabs.logging.core.statistics.datatypes.StringTracker;
import com.whylabs.logging.core.types.TypedDataConverter;
import java.util.Set;
import java.util.stream.Stream;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.val;

@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Getter
@Builder(setterPrefix = "set")
public class ColumnProfile {
  private static final Set<Type> NUMERIC_TYPES =
      Stream.of(Type.FRACTIONAL, Type.INTEGRAL).collect(toSet());

  final String columnName;
  final CountersTracker counters;
  final SchemaTracker schemaTracker;
  final NumberTracker numberTracker;
  final StringTracker stringTracker;

  public ColumnProfile(String columnName) {
    this.columnName = columnName;
    this.counters = new CountersTracker();
    this.schemaTracker = new SchemaTracker();
    this.numberTracker = new NumberTracker();
    this.stringTracker = new StringTracker();
  }

  public void track(Object value) {
    synchronized (this) {
      counters.incrementCount();

      if (value == null) {
        counters.incrementNull();
        return;
      }

      // always track text information
      // TODO: ignore this if we already know the data type
      if (value instanceof String) {
        stringTracker.update((String) value);
      }

      val typedData = TypedDataConverter.convert(value);
      schemaTracker.track(typedData.getType());

      switch (typedData.getType()) {
        case FRACTIONAL:
          numberTracker.track(typedData.getFractional());
          break;
        case INTEGRAL:
          numberTracker.track(typedData.getIntegralValue());
          break;
        case BOOLEAN:
          if (typedData.isBooleanValue()) {
            counters.incrementTrue();
          }
          break;
      }
    }
  }

  public ColumnSummary toColumnSummary() {
    val schema = fromSchemaTracker(schemaTracker);

    val builder = ColumnSummary.newBuilder().setCounters(counters.toProtobuf());
    if (schema != null) {
      builder.setSchema(schema);
      if (schema.getInferredType().getType() == Type.STRING) {
        val stringSummary = SummaryConverters.fromStringTracker(stringTracker);
        if (stringSummary != null) {
          builder.setStringSummary(stringSummary);
        }
      } else if (NUMERIC_TYPES.contains(schema.getInferredType().getType())) {
        val numberSummary = SummaryConverters.fromNumberTracker(this.numberTracker);
        if (numberSummary != null) {
          builder.setNumberSummary(numberSummary);
        }
      }
    }

    return builder.build();
  }

  public ColumnMessage.Builder toProtobuf() {
    return ColumnMessage.newBuilder()
        .setName(columnName)
        .setCounters(counters.toProtobuf())
        .setSchema(schemaTracker.toProtobuf())
        .setNumbers(numberTracker.toProtobuf())
        .setStrings(stringTracker.toProtobuf());
  }

  public static ColumnProfile fromProtobuf(ColumnMessage message) {
    return ColumnProfile.builder()
        .setColumnName(message.getName())
        .setCounters(CountersTracker.fromProtobuf(message.getCounters()))
        .setSchemaTracker(SchemaTracker.fromProtobuf(message.getSchema()))
        .setNumberTracker(NumberTracker.fromProtobuf(message.getNumbers()))
        .build();
  }
}

File: whylabs/logging/core/columnprofile.py
#!/usr/bin/env python3
"""
created 5/7/20 by ibackus
"""
from whylabs.logging.core.types import TypedDataConverter
from whylabs.logging.core.statistics import NumberTracker
from whylabs.logging.core.summaryconverters import from_number_tracker
from whylabs.logging.core.data import ColumnSummary, ColumnMessage


class ColumnProfile:
    """
    Parameters
    ----------
    name : str
        Name of the column profile

    TODO:
        * _NUMERIC_TYPES for number type checking
        * CountersTracker
        * SchemaTracker
        * StringTracker
        * Proper TypedDataConverter type checking
        * Multi-threading/parallelism
    """
    _NUMERIC_TYPES = None  # TODO: Implement

    def __init__(self, name, number_tracker=None):
        self.column_name = name
        # self.counters = CountersTracker()
        # self.schemaTracker = SchemaTracker()
        if number_tracker is None:
            number_tracker = NumberTracker()
        self.number_tracker = number_tracker
        # self.stringTracker = StringTracker()

    def track(self, value):
        typed_data = TypedDataConverter.convert(value)
        if isinstance(typed_data, (float, int)):
            self.number_tracker.track(typed_data)
        elif isinstance(typed_data, bool):
            # TODO: implement boolean counter
            ...

    def to_summary(self):
        schema = 123
        # TODO: implement the real schema/type checking
        opts = {}
        if schema is not None:
            if isinstance(schema, str):
                ...
            elif isinstance(schema, (float, int)):
                numbersSummary = from_number_tracker(self.number_tracker)
                if numbersSummary is not None:
                    opts['number_summary'] = numbersSummary
        return ColumnSummary(**opts)

    def to_protobuf(self):
        return ColumnMessage(
            name=self.column_name,
            # counters=self.counters.toProtonbuf(),
            # schema=self.schemaTracker.to_protobuf(),
            numbers=self.number_tracker.to_protobuf(),
            # strings=self.stringTracker.to_protobuf(),
        )

    @staticmethod
    def from_protobuf(message):
        return ColumnProfile(
            message.name,
            # counters=<class>.from_protobuf(),
            # schemaTracker=<class>.from_protobuf(),
            numberTracker=NumberTracker.from_protobuf(message.numbers),
        )
